relayer-engine/src/storage/storage.ts

```diff
@@ -86,6 +104,88 @@ export class DefaultStorage implements Storage {
     this.plugins = new Map(plugins.map(p => [p.pluginName, p]));
   }
 
+  // fetch an emitter record by chainId and emitterAddress
+  getEmitterRecord(
+    pluginName: string,
+    chainId: ChainId,
+    emitterAddress: string,
+  ): Promise<EmitterRecord | null> {
+    return this.store.withRedis(async redis => {
+      return await this.getEmitterRecordInner(
+        redis,
+        emitterRecordKey(pluginName, chainId, emitterAddress),
+      );
+    });
+  }
+
+  private async getEmitterRecordInner(
+    redis: IRedis,
+    key: string,
+  ): Promise<EmitterRecord | null> {
+    const res = await redis.hGet(EMITTER_KEY, key);
+    if (!res) {
+      return null;
+    }
+    const parsed: EmitterRecord = JSON.parse(res);
+    parsed.time = new Date(parsed.time);
+    return parsed;
+  }
+
+  // set an emitter record with given sequence and update the timestamp
+  setEmitterRecord(
+    pluginName: string,
+    chainId: ChainId,
+    emitterAddress: string,
+    lastSeenSequence: number,
+  ): Promise<void> {
+    return this.store.runOpWithRetry(async redis => {
+      this.logger.debug(`setEmitterRecord`);
+      while (true) {
+        const record: EmitterRecord = { lastSeenSequence, time: new Date() };
+        const key = emitterRecordKey(pluginName, chainId, emitterAddress);
+
+        const entry = await this.getEmitterRecordInner(redis, key);
```

Comment:
having an "`Inner`" function that accepts an IRedis solves the problem of double `executeIsolated` calls, but this still feels dangerous

Reply:
Needed to do this for aquireLockUnsafe and releaseLockUnsafe too

Reply:
it's private so I guess it's ok, it's like passing around a `tx` object when you're using a db. Maybe what we could do is for functions to receive an optional argument at the end which is redis. If it's passed in, we use it, else we call executeIsolated. If this were a convention it would feel less bad maybe?

Reply:
That's what I did initially actually, but deadlocking the process is such a footgun that I made the arg required. This seems like something that could always be inside an existing "tx"